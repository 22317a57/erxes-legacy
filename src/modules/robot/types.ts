export type IFeature = {
  name: string;
  text: string;
};
