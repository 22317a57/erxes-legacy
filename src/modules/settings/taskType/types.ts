export interface ITaskType {
  _id: string;
  name: string;
  icon: string;
}

// query types

export type TaskTypeQueryResponse = {
  taskTypes: ITaskType[];
  loading: boolean;
  refetch: () => void;
};

export type MutationVariables = {
  _id?: string;
  name?: string;
  icon?: string;
};

// mutation types

export type AddMutationResponse = {
  addMutation: (mutation: { variables: MutationVariables }) => Promise<any>;
};

export type EditMutationResponse = {
  editMutation: (mutation: { variables: MutationVariables }) => Promise<any>;
};

export type RemoveMutationResponse = {
  removeMutation: (mutation: { variables: { _id: string } }) => Promise<any>;
};
