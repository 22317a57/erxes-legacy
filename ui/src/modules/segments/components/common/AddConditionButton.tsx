import Button from 'modules/common/components/Button';
import Icon from 'modules/common/components/Icon';
import { __ } from 'modules/common/utils';
import React from 'react';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Popover from 'react-bootstrap/Popover';
import { ISegmentCondition, ISegmentField } from '../../types';
import { FieldType, PopoverList } from '../styles';

type Props = {
  fields: ISegmentField[];
  events: string[];
  addCondition: (condition: ISegmentCondition) => void;
};

class AddConditionButton extends React.Component<Props> {
  private overlayTrigger;

  addCondition = (type: string) => {
    this.props.addCondition({
      key: Math.random().toString(),
      propertyName: '',
      propertyValue: '',
      propertyOperator: '',
      type,
    });

    this.overlayTrigger.hide();
  };

  renderPopover() {
    const addPropertyCondition = () => {
      return this.addCondition('property');
    }

    const addEventCondition = () => {
      return this.addCondition('event');
    }

    return (
      <Popover id="condition-popover">
        <Popover.Title as="h3">{__('Select a field')}</Popover.Title>
        <Popover.Content>
          <PopoverList onClick={addPropertyCondition}>
            <FieldType>
              {__("Fields")}
              <Icon icon="information" />
            </FieldType>
          </PopoverList>

          <PopoverList onClick={addEventCondition}>
            <FieldType>
              {__("Events")}
              <Icon icon="settings-3" />
            </FieldType>
          </PopoverList>
        </Popover.Content>
      </Popover>
    );
  }

  render() {
    return (
      <OverlayTrigger
        ref={overlayTrigger => {
          this.overlayTrigger = overlayTrigger;
        }}
        trigger="click"
        placement="bottom"
        overlay={this.renderPopover()}
        container={this}
        rootClose={true}
      >
        <Button btnStyle="success" icon="add">
          Add a condition
        </Button>
      </OverlayTrigger>
    );
  }
}

export default AddConditionButton;
