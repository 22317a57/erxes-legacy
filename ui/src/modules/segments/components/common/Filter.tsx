import Button from "modules/common/components/Button";
import { FormControl } from "modules/common/components/form";
import { __ } from "modules/common/utils";
import { operators } from "modules/customers/constants";
import { FlexContent, FlexItem, FlexRightItem } from "modules/layout/styles";
import React from "react";
import { IConditionFilter } from "../../types";

type Props = {
  names: string[];
  filter: IConditionFilter;
  onChange: (filter: IConditionFilter) => void;
  onRemove?: (id: string) => void;
};

type State = {
  key: string;
  currentName: string;
  currentOperator: string;
  currentValue: string;
};

class Filter extends React.Component<Props, State> {
  constructor(props) {
    super(props);

    const { filter } = this.props;

    this.state = {
      key: filter.key || '',
      currentName: filter.name,
      currentOperator: filter.operator,
      currentValue: filter.value,
    };
  }

  onChange = () => {
    const { currentName, currentOperator, currentValue } = this.state;
    const { onChange, filter } = this.props;

    return onChange({
      key: filter.key,
      name: currentName,
      operator: currentOperator,
      value: currentValue,
    });
  }

  onChangeValue = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentValue: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  onChangeNames = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentName: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  onChangeOperators = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentOperator: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  renderNames() {
    const { names } = this.props;
    const { currentName } = this.state;

    return (
      <FormControl componentClass="select" placeholder={__("select")} onChange={this.onChangeNames} value={currentName}>
        <option />
        {names.map((name, index) => (
          <option value={name} key={index}>
            {name}
          </option>
        ))}
      </FormControl>
    );
  }

  renderOperators() {
    const { currentOperator } = this.state;

    return (
      <FormControl componentClass="select" placeholder={__("select")} onChange={this.onChangeOperators} value={currentOperator}>
        <option />
        {operators.map(c => (
          <option value={c.value} key={c.value}>
            {c.name}
          </option>
        ))}
      </FormControl>
    );
  }

  onRemove = () => {
    const { onRemove } = this.props;

    if (onRemove) {
      onRemove(this.props.filter.key || '');
    }
  };

  renderRemoveButton = () => {
    const { onRemove } = this.props;

    if (!onRemove) {
      return;
    }

    return <Button btnStyle="danger" size="small" icon="cancel-1" onClick={this.onRemove} />;
  }

  render() {
    const { currentValue } = this.state;

    return (
      <FlexContent>
        <FlexItem>
          {this.renderNames()}
          {this.renderOperators()}
          <input value={currentValue} onChange={this.onChangeValue} />
        </FlexItem>
        <FlexRightItem>
          {this.renderRemoveButton()}
        </FlexRightItem>
      </FlexContent>
    );
  }
}

export default Filter;
