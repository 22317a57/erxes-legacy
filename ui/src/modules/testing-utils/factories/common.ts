import * as Factory from 'factory.ts';
import { IConditionsRule } from 'modules/common/types';

export const conditionsRuleFactory = Factory.Sync.makeFactory<IConditionsRule>({
  _id: '1',
  kind: 'string',
  text: 'string',
  condition: 'string',
  value: 'string',
});