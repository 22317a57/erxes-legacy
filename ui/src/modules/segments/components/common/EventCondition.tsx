import Button from 'modules/common/components/Button';
import { FormControl } from 'modules/common/components/form';
import { __ } from 'modules/common/utils';
import { FlexContent, FlexItem, FlexRightItem } from 'modules/layout/styles';
import React from 'react';
import { IConditionFilter } from '../../types';
import { ConditionItem } from '../styles';
import Filter from './Filter';

type Props = {
  events: string[];
  conditionKey: string;
  name: string;
  attributeNames: string[];
  attributeFilters: IConditionFilter[];
  onChange: (args: { key: string, name: string, attributeFilters: IConditionFilter[] }) => void;
  onRemove: (id: string) => void;
};

type State = {
  attributeFilters: IConditionFilter[];
  currentEvent?: string;
}

class Condition extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      currentEvent: props.name,
      attributeFilters: props.attributeFilters.map(filter => ({ key: Math.random().toString(), ...filter }))
    }
  }

  removeCondition = () => {
    this.props.onRemove(this.props.conditionKey);
  };

  onChangeFilter = () => {
    const { onChange, conditionKey } = this.props;
    const { currentEvent, attributeFilters } = this.state;

    return onChange({
      key: conditionKey,
      name: currentEvent || '',
      attributeFilters,
    });
  }

  onChangeAttributeFilter = (filter: IConditionFilter) => {
    this.setState({
      attributeFilters: this.state.attributeFilters.map(f =>
        f.key === filter.key ? filter : f
      )
    }, this.onChangeFilter);
  };

  onRemoveAttributeFilter = (key: string) => {
    const attributeFilters = this.state.attributeFilters.filter(f => f.key !== key);

    this.setState({ attributeFilters }, this.onChangeFilter);
  };

  onChangeEvents = (e) => {
    this.setState({ currentEvent: e.currentTarget.value }, this.onChangeFilter);
  }

  addAttributeFilter = () => {
    const attributeFilter = {
      key: Math.random().toString(),
      name: '',
      operator: '',
      value: ''
    };

    const { attributeFilters } = this.state;

    attributeFilters.push(attributeFilter);

    this.setState({ attributeFilters });
  }

  renderAttributeFilters = () => {
    const { attributeFilters } = this.state;
    const { attributeNames } = this.props;

    return attributeFilters.map((filter, index) => {
      return (
        <Filter
          key={index}
          names={attributeNames}
          filter={filter}
          onChange={this.onChangeAttributeFilter}
          onRemove={this.onRemoveAttributeFilter}
        />
      );
    });
  }

  renderNames() {
    const { events } = this.props;
    const { currentEvent } = this.state;

    return (
      <FormControl componentClass="select" placeholder={__("select")} onChange={this.onChangeEvents} value={currentEvent}>
        <option />

        {events.map((name, index) => (
          <option value={name} key={index}>
            {name}
          </option>
        ))}
      </FormControl>
    );
  }

  render() {
    return (
      <ConditionItem>
        <FlexContent>
          <FlexItem>
            {this.renderNames()}

            <div>
              {this.renderAttributeFilters()}
            </div>

            <div>
              <Button onClick={this.addAttributeFilter}>
                Add attribute
              </Button>
            </div>
          </FlexItem>

          <FlexRightItem>
            <Button
              btnStyle="danger"
              size="small"
              icon="cancel-1"
              onClick={this.removeCondition}
            />
          </FlexRightItem>
        </FlexContent>
      </ConditionItem>
    );
  }
}

export default Condition;
