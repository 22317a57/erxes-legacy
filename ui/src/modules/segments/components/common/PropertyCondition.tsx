import Button from 'modules/common/components/Button';
import { FlexContent, FlexItem, FlexRightItem } from 'modules/layout/styles';
import React from 'react';
import { IConditionFilter } from '../../types';
import { ConditionItem } from '../styles';
import Filter from './Filter';

type Props = {
  fields: any[];
  conditionKey: string;
  name: string,
  operator: string,
  value: string,
  onChange: (args: { key: string, name: string, operator: string, value: string }) => void;
  onRemove: (id: string) => void;
};

class Condition extends React.Component<Props, {}> {
  removeCondition = () => {
    this.props.onRemove(this.props.conditionKey);
  };

  onChangeFilter = (filter: IConditionFilter) => {
    const { onChange, conditionKey } = this.props;

    return onChange({
      key: conditionKey,
      name: filter.name,
      operator: filter.operator,
      value: filter.value
    });
  }

  render() {
    const { fields, conditionKey, name, operator, value } = this.props;
    const names = fields.map(field => field.name);

    const filter = {
      key: conditionKey,
      name,
      operator,
      value
    }

    return (
      <ConditionItem>
        <FlexContent>
          <FlexItem>
            <Filter names={names} filter={filter} onChange={this.onChangeFilter} />
          </FlexItem>
          <FlexRightItem>
            <Button
              btnStyle="danger"
              size="small"
              icon="cancel-1"
              onClick={this.removeCondition}
            />
          </FlexRightItem>
        </FlexContent>
      </ConditionItem>
    );
  }
}

export default Condition;
