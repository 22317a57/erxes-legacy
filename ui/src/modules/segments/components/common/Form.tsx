import Button from 'modules/common/components/Button';
import FormControl from 'modules/common/components/form/Control';
import CommonForm from 'modules/common/components/form/Form';
import FormGroup from 'modules/common/components/form/Group';
import ControlLabel from 'modules/common/components/form/Label';
import { IButtonMutateProps, IFormProps } from 'modules/common/types';
import { __, generateRandomColorCode } from 'modules/common/utils';
import { FlexContent, FlexItem } from 'modules/layout/styles';
import {
  IConditionFilter,
  IEvent,
  ISegment,
  ISegmentCondition,
  ISegmentWithConditionDoc
} from 'modules/segments/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { ConditionWrapper, SegmentTitle, SegmentWrapper } from '../styles';
import AddConditionButton from './AddConditionButton';
import EventCondition from './EventCondition';
import PropertyCondition from './PropertyCondition';

type Props = {
  contentType?: string;
  fields: any[];
  events: IEvent[];
  renderButton: (props: IButtonMutateProps) => JSX.Element;
  edit?: (params: { _id: string; doc: ISegmentWithConditionDoc }) => void;
  segment?: ISegment;
  headSegments: ISegment[];
  count: (segment: ISegment) => void;
  isForm?: boolean;
  afterSave?: () => void;
};

type State = {
  name: string;
  description: string;
  subOf: string;
  color: string;
  conditions: ISegmentCondition[];
};

class Form extends React.Component<Props, State> {
  constructor(props) {
    super(props);

    const segment: ISegment = props.segment || {
      name: '',
      description: '',
      subOf: '',
      color: generateRandomColorCode(),
      conditions: [],
    };

    segment.conditions = segment.conditions.map(
      (cond: ISegmentCondition) => ({
        key: Math.random().toString(),
        ...cond
      })
    );

    this.state = segment;
  }

  addCondition = (condition: ISegmentCondition) => {
    this.setState({
      conditions: [...this.state.conditions, condition]
    });
  };

  changeEventCondition = (args: { key: string, name: string, attributeFilters: IConditionFilter[] }) => {
    const condition = {
      type: 'event',
      key: args.key,
      eventName: args.name,
      eventAttributeFilters: (args.attributeFilters || []).map(filter => {
        const { key, ...rest } = filter;

        return rest;
      })
    }

    this.setState({
      conditions: this.state.conditions.map(c =>
        c.key === condition.key ? condition : c
      )
    });
  };

  changePropertyCondition = (args: { key: string, name: string, operator: string, value: string }) => {
    const condition = {
      type: 'property',
      key: args.key,
      propertyName: args.name,
      propertyOperator: args.operator,
      propertyValue: args.value,
    }

    this.setState({
      conditions: this.state.conditions.map(c =>
        c.key === condition.key ? condition : c
      )
    });
  };

  removeCondition = (key: string) => {
    const conditions = this.state.conditions.filter(c => c.key !== key);

    this.setState({ conditions });
  };

  handleChange = <T extends keyof State>(name: T, value: State[T]) => {
    this.setState({ [name]: value } as Pick<State, keyof State>);
  };

  generateDoc = (values: {
    _id?: string;
    name: string;
    subOf: string;
    color: string;
  }) => {
    const { segment, contentType } = this.props;
    const { conditions } = this.state;
    const finalValues = values;

    const updatedConditions: ISegmentCondition[] = [];

    if (segment) {
      finalValues._id = segment._id;
    }

    conditions.forEach((cond: ISegmentCondition) => {
      const { key, ...rest } = cond;
      updatedConditions.push(rest);
    });

    return {
      ...finalValues,
      contentType,
      conditions: updatedConditions
    };
  };

  renderParent() {
    const { contentType } = this.props;
    const { subOf } = this.state;

    if (!subOf) {
      return null;
    }

    return (
      <React.Fragment>
        <Link
          to={`/segments/edit/${contentType}/${subOf}`}
          target="_blank"
        >
          <Button icon="eye" ignoreTrans={true}>
            {__('Parent segment conditions')}
          </Button>
        </Link>
        <hr />
      </React.Fragment>
    );
  }

  renderCondition(condition: ISegmentCondition) {
    const { fields, events } = this.props;

    if (condition.type === 'property') {
      return (
        <PropertyCondition
          fields={fields}
          key={condition.key}
          conditionKey={condition.key || ''}
          name={condition.propertyName || ''}
          operator={condition.propertyOperator || ''}
          value={condition.propertyValue || ''}
          onChange={this.changePropertyCondition}
          onRemove={this.removeCondition}
        />
      )
    }

    return (
      <EventCondition
        events={events}
        key={condition.key}
        conditionKey={condition.key || ''}
        name={condition.eventName || ''}
        attributeFilters={condition.eventAttributeFilters || []}
        onChange={this.changeEventCondition}
        onRemove={this.removeCondition}
      />
    )
  }

  renderConditions() {
    const { conditions } = this.state;

    return (
      <React.Fragment>
        <ConditionWrapper>
          {this.renderParent()}
          {conditions.map(condition => this.renderCondition(condition))}
          </ConditionWrapper>
        <AddConditionButton addCondition={this.addCondition} />
      </React.Fragment>
    );
  }

  renderSubOf(formProps: IFormProps) {
    const onChange = (e: React.FormEvent) =>
      this.handleChange('subOf', (e.currentTarget as HTMLInputElement).value);

    return (
      <FormGroup>
        <ControlLabel>Sub segment of</ControlLabel>
        <FormControl
          {...formProps}
          name="subOf"
          componentClass="select"
          value={this.state.subOf || ''}
          onChange={onChange}
        >
          <option value="">[not selected]</option>
          {this.props.headSegments.map(segment => (
            <option value={segment._id} key={segment._id}>
              {segment.name}
            </option>
          ))}
        </FormControl>
      </FormGroup>
    );
  }

  renderForm = (formProps: IFormProps) => {
    const {
      isForm,
      segment,
      contentType,
      renderButton,
      afterSave
    } = this.props;

    const { values, isSubmitted } = formProps;
    const { name, description, color } = this.state;

    const nameOnChange = (e: React.FormEvent) =>
      this.handleChange('name', (e.currentTarget as HTMLInputElement).value);

    const descOnChange = (e: React.FormEvent) =>
      this.handleChange(
        'description',
        (e.currentTarget as HTMLInputElement).value
      );

    const colorOnChange = (e: React.FormEvent) =>
      this.handleChange('color', (e.currentTarget as HTMLInputElement).value);

    return (
      <>
        <FormGroup>
          <ControlLabel required={true}>Name</ControlLabel>
          <FormControl
            {...formProps}
            name="name"
            value={name}
            onChange={nameOnChange}
            required={true}
            autoFocus={true}
          />
        </FormGroup>
        <FormGroup>
          <ControlLabel>Description</ControlLabel>
          <FormControl
            {...formProps}
            name="description"
            value={description}
            onChange={descOnChange}
          />
        </FormGroup>
        {this.renderSubOf({ ...formProps })}
        <FormGroup>
          <ControlLabel>Color</ControlLabel>
          <FormControl
            {...formProps}
            name="color"
            type="color"
            value={color}
            onChange={colorOnChange}
          />
        </FormGroup>
        <Button.Group>
          {isForm && (
            <Link to={`/segments/${contentType}`}>
              <Button size="small" btnStyle="simple" icon="cancel-1">
                Cancel
              </Button>
            </Link>
          )}

          {renderButton({
            name: 'segment',
            values: this.generateDoc(values),
            callback: afterSave,
            isSubmitted,
            object: segment
          })}
        </Button.Group>
      </>
    );
  };

  render() {
    return (
      <SegmentWrapper>
        <SegmentTitle>{__('Filters')}</SegmentTitle>
        {this.renderConditions()}
        <hr />
        <FlexContent>
          <FlexItem count={4}>
            <CommonForm renderContent={this.renderForm} />
          </FlexItem>
          <FlexItem count={2} />
        </FlexContent>
      </SegmentWrapper>
    );
  }
}

export default Form;
