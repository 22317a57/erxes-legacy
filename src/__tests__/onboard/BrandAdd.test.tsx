import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { IButtonMutateProps } from '../../modules/common/types';
import BrandAdd from '../../modules/onboard/components/BrandAdd';

describe('BrandAdd component', () => {
  const defaultProps = {
    brandsTotalCount: 0,
    changeStep: () => null,
    renderButton: (props: IButtonMutateProps) => <div />
  };

  test('renders BrandAdd successfully', () => {
    shallow(<BrandAdd {...defaultProps} />);
  });

  test('renders successfully with default value', () => {
    const wrapper = mount(<BrandAdd {...defaultProps} />);
    const props = wrapper.props();

    expect(props).toMatchObject(defaultProps);
  });

  test('snapshot matches', () => {
    const rendered = renderer.create(<BrandAdd {...defaultProps} />).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
