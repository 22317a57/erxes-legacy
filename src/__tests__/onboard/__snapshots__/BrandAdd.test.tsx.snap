// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrandAdd component snapshot matches 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="sc-dqBHgY iDAfER"
  >
    <h2>
      Let's create your brand
    </h2>
    <div
      className="sc-jKVCRD eDSGVr"
    >
      <label
        className="sc-ebFjAB bHuvQk"
      >
        Brand Name
        <span>
           *
        </span>
      </label>
      <input
        autoFocus={true}
        className="sc-kaNhvL bKDVoH"
        disabled={false}
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="sc-jtRfpW ksjgFM"
  >
    <div>
      <button
        className="sc-ifAKCX bKlEue"
        disabled={true}
        type="button"
      >
        Previous
      </button>
      <button
        className="sc-ifAKCX cQdFDO"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Next
         
        <i
          className="icon-rightarrow-2 sc-bwzfXH hpxbGw"
          icon="rightarrow-2"
        />
      </button>
    </div>
    <a
      onClick={[Function]}
    >
      Skip for now
       Â»
    </a>
  </div>
</form>
`;
